### ISSUE TITLE
SPIR-V validation error: "All OpVariable instructions in a function must be the first instructions in the first block."

### ISSUE THREAD

**Issue by karolherbst** [intel/llvm#9957]
Multiple SyCL CTS tests are triggering the "All OpVariable instructions in a function must be the first instructions in the first block." spir-v validation error, e.g. `test_accessor`.

This appears to mostly happen on function calls e.g.:

```
      %entry = OpLabel
       %2015 = OpFunctionCall %void %__itt_offload_wi_start_wrapper
 %data_i_i_i = OpVariable %_ptr_Function_class_sycl___V1__vec_837 Function
```

From the SPIR-V specification "2.4. Logical Layout of a Module"

All [OpVariable](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpVariable) instructions in a function must be in the first block in the function. These instructions, together with any intermixed [OpLine](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpLine) and [OpNoLine](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpNoLine) instructions, must be the first instructions in that block. (Note the validation rules prevent [OpPhi](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpPhi) instructions in the first block of a function.)

**Comment by vmaksimo** [comment]
should be fixed by https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/2439
need to verify after the change reaches intel/llvm

**Comment by MrSidims** [comment]
> should be fixed by [KhronosGroup/SPIRV-LLVM-Translator#2439](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/2439) need to verify after the change reaches intel/llvm

Not exactly, it fixes some cases, but the patch doesn't move alloca/OpVariable from not-the-first basic block to the entry of the function.

**Comment by maarquitos14** [comment]
@karolherbst I have verified locally and I cannot reproduce this issue anymore. Can you please verify on your end? Thanks in advance for your help :)

**Comment by karolherbst** [comment]
I've checked today and this issue _seems_ to be fixed. It's kinda hard to tell because I'm also running into other validation bugs. But anyway, this one seems to be fine now, thanks.

If I see it happening again I'll just report back.
